utils/tracefs.c

```diff
@@ -1,66 +1,44 @@
 #include <fcntl.h>
 #include <stdio.h>
 #include <string.h>
+#include <sys/vfs.h>
 
 #include "utils/tracefs.h"
 #include "utils/utils.h"
 
 #define PROC_MOUNTINFO "/proc/self/mountinfo"
 
+#define TRACEFS_DIR_PATH "/sys/kernel/tracing"
+#define TRACEFS_MAGIC 0x74726163
+
+#define DEBUGFS_DIR_PATH "/sys/kernel/debug/tracing"
+#define DEBUGFS_MAGIC 0x64626720
+
 static char *TRACING_DIR = NULL;
 
 static bool find_tracing_dir(void)
 {
-	FILE *fp;
-	char *line = NULL, fs_type[NAME_MAX], mount_point[PATH_MAX];
-	static char debugfs_suffix[] = "tracing";
-	bool debugfs_found = false;
-	size_t len;
-
-	if (TRACING_DIR)
-		return false;
-
-	fp = fopen(PROC_MOUNTINFO, "r");
-	if (fp == NULL)
-		return false;
-
-	while (getline(&line, &len, fp) > 0) {
-		/*
-		 * /proc/<pid>/mountinfo format:
-		 * 36 35 98:0 /mnt1 /mnt2 rw,noatime master:1 - ext3 .... ....
-		 * (1)(2)(3)   (4)   (5)      (6)      (7)   (8) (9) (10) (11)
-		 *                mount_point                  fs_type
-		 *
-		 * (9) is the file system type, (5) is the mount point relative
-		 * to self's root directory.
-		 */
-		sscanf(line, "%*i %*i %*u:%*u %*s %s %*s %*s - %s %*s %*s\n", mount_point, fs_type);
-
-		if (!strcmp(fs_type, "tracefs")) {
-			/* discard previously kept debugfs tracing dir */
-			if (TRACING_DIR)
-				free(TRACING_DIR);
-			xasprintf(&TRACING_DIR, "%s", mount_point);
-			pr_dbg2("Found tracefs at %s\n", mount_point);
-			pr_dbg2("Use %s as TRACING_DIR\n", TRACING_DIR);
-			return true;
-		}
-
-		if (!strcmp(fs_type, "debugfs")) {
-			xasprintf(&TRACING_DIR, "%s/%s", mount_point, debugfs_suffix);
-			pr_dbg2("Found debugfs at %s\n", mount_point);
-			pr_dbg2("Keep searching for tracefs...\n");
-			debugfs_found = true;
-		}
+	struct statfs fs;
+
+	/* check tracefs path is mounted */
+	statfs(TRACEFS_DIR_PATH, &fs);
+	if (fs.f_type == TRACEFS_MAGIC) {
+		xasprintf(&TRACING_DIR, "%s", TRACEFS_DIR_PATH);
+		pr_dbg2("TRACING_DIR : %s\n", TRACING_DIR);
+
+		return true;
 	}
 
-	/* we couldn't find a tracefs, but found a debugfs... */
-	if (debugfs_found) {
-		pr_dbg2("Use %s as TRACING_DIR\n", TRACING_DIR);
+	/* check debugfs path is mounted */
+	statfs(DEBUGFS_DIR_PATH, &fs);
+	if (fs.f_type == DEBUGFS_MAGIC) {
```

Comment:
This check actually fails because `/sys/kernel/debug` is debugfs, but  `/sys/kernel/debug/tracing` is mounted as tracefs on top of the debugfs.

There might be a chance that old kernel uses `/sys/kernel/debug/tracing` as debugfs so you might be better to check both `DEBUGFS_MAGIC` and `TRACEFS_MAGIC`.